client/config/config.go

```diff
@@ -0,0 +1,234 @@
+package config
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+
+	toml "github.com/pelletier/go-toml"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+)
+
+const (
+	flagGet = "get"
+)
+
+/*
+
+Plan:
+
+0. When the user uses the gaiacli config node command to set the node value,
+ the new value is written successfully to the ~/.gaiacli/config/config.toml
+
+1.Create $NODE_DIR/config/client.toml to hold all client-side configuration
+(e.g. keyring-backend, node, chain-id etc...)+
+
+
+2. Enable get functionality of client-side configuration
+
+get examples
+
+config keyring-backend --get //keyring-backend returns value from config file
+config with no args --get //prints default config and exits
+config invalidArg --get // returns error unknown conf key
+config keyring-backend output --get // returns error too many arguments
+
+config without arguments // just print config from config file
+
+
+3.Enable set functionality of client side configuration
+
+config keyring-backend test // sets keyring-backend = test - expected behavior
+
+config keyring-backend with no arguments // outputs default
+config invalidKey invalidValue
+
+
+CLI flags
+
+do i need to add all flags? from client/flags/flags.go
+
+"keyring-dir" : "",
+"ledger": false,
+"height": 0,
+"gas-adjustment": flags.DefaultGasAdjustment,//float
+"from": "",
+"account-number": 0,
+"sequence": 0,
+"memo": "",
+"fees": "",
+"gas-prices":"",
+"dry-run": false,
+"generate-only":false,
+"offline": false,
+"yes": true, //doublec heck skip-confirmation = true
+
+
+
+
+
+
+
+
+*/
+
+var ErrWrongNumberOfArgs = fmt.Errorf("wrong number of arguments")
+
+// is this full client side configuration?
+// add height = 0
+var configDefaults = map[string]string{
+
+	"chain-id":        "",
+	"keyring-backend": "os",
+	"output":          "text",
+	"node":            "tcp://localhost:26657",
+	"broadcast-mode":  "sync",
+}
+
+// ConfigCmd returns a CLI command to interactively create an application CLI
+// config file.
+func Cmd(defaultCLIHome string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "config <key> [value]",
+		Short: "Create or query an application CLI configuration file",
+		RunE:  runConfigCmd,
+		Args:  cobra.RangeArgs(0, 2),
+	}
+
+	cmd.Flags().String(flags.FlagHome, defaultCLIHome,
+		"set client's home directory for configuration")
+	cmd.Flags().Bool(flagGet, false,
+		"print configuration value or its default if unset")
+	return cmd
+}
+
+func runConfigCmd(cmd *cobra.Command, args []string) error {
+	cfgFile, err := ensureConfFile(viper.GetString(flags.FlagHome))
+	if err != nil {
+		return err
+	}
+
+	getAction := viper.GetBool(flagGet)
```

Comment:
In theory, one could even make this design simpler, i.e.:

1. if _zero_ args are passed, then it's a `get-all`, i.e. print to `/dev/stdout` all set configuration values (and fallback to defaults for those which are unset)
2. if _only one_ arg is passed, then it's a `get`
3. if _two args_ are passed, then it's a `set`
4. exit with 1 and print an error to `/dev/stderr` in all other cases.

Wdyt?

Reply:
The benefits of the suggested approach is that we avoid handling the `--get` flag and the result is more user-friendly.

My 2c

Reply:
I like this approach. Seems better for UX.

Reply:
Also, step 2 in #8529 is removing dependency on the global `viper` instance. If we can not introduce dependency on that singleton, it might help step 2.

Reply:
> Also, step 2 in #8529 is removing dependency on the global `viper` instance. If we can not introduce dependency on that singleton, it might help step 2.

you mean to remove global  viper instance in whole cosmos-sdk repositoty?

Reply:
> you mean to remove global viper instance in whole cosmos-sdk repositoty?

yeah, so in step 2: remove `viper.Get/Set`, and use scoped instances: `ctx.Viper.Get/Set`.

For this PR, I propose to not introduce new `viper.Get/Set`.

Reply:
> In theory, one could even make this design simpler, i.e.:
> 
> 1. if _zero_ args are passed, then it's a `get-all`, i.e. print to `/dev/stdout` all set configuration values (and fallback to defaults for those which are unset)
> 2. if _only one_ arg is passed, then it's a `get`
> 3. if _two args_ are passed, then it's a `set`
> 4. exit with 1 and print an error to `/dev/stderr` in all other cases.
> 
> Wdyt?

@alessio  thanks for your guidance. Following @amaurym  suggestion, we decided to stick with toml config [file](https://github.com/cyberbono3/cosmos-sdk/blob/cyberbono3/8529-add-client-config/client/config/toml.go#L13-L27) as it is implemented in tendermint [config](https://github.com/cyberbono3/cosmos-sdk/blob/cyberbono3/8529-add-client-config/server/config/toml.go#L11). So I  use [InitConfigTemplate](https://github.com/cyberbono3/cosmos-sdk/blob/cyberbono3/8529-add-client-config/client/config/toml.go#L31) to generate config template and then write it to the client.toml file using [WriteConfigFile](https://github.com/cyberbono3/cosmos-sdk/blob/cyberbono3/8529-add-client-config/client/config/toml.go#L54).

Reply:
> > you mean to remove global viper instance in whole cosmos-sdk repositoty?
> 
> yeah, so in step 2: remove `viper.Get/Set`, and use scoped instances: `ctx.Viper.Get/Set`.
> 
> For this PR, I propose to not introduce new `viper.Get/Set`.

For this PR I have to get viper instance from ctx.Viper.Get/Set and get rid of local viper [instance](https://github.com/cyberbono3/cosmos-sdk/blob/cyberbono3/8529-add-client-config/client/config/config.go#L58) right?

In the step 2 , I start a new PR from scratch and get rid of all `viper.Get/Set`  and use scoped instances: `ctx.Viper.Get/Set` in whole cosmos-sdk repository. Does it sound good?

Reply:
> > you mean to remove global viper instance in whole cosmos-sdk repositoty?
> 
> yeah, so in step 2: remove `viper.Get/Set`, and use scoped instances: `ctx.Viper.Get/Set`.
> 
> For this PR, I propose to not introduce new `viper.Get/Set`.

I would like to double check. Under scope instance you mean this [context](https://github.com/cosmos/cosmos-sdk/blob/master/server/util.go#L55), right? So I have to use Context struct with Viper and Config fields in client/config.go , right?

Reply:
I was actually thinking about the clientCtx, [this one](https://github.com/cosmos/cosmos-sdk/blob/280ee4f15e46117a5061ab3a374d9b695ed082bf/client/context.go#L22). We maybe don't need an additional scoped `viper` field in that struct.

@alessio of @alexanderbez, do you have any advice on this? i.e how do we get a `ClientConfig{}` struct to be populated into `client.Context`? (using viper or not)